### PR TITLE
raidboss: p3n trigger

### PR DESCRIPTION
**Pull request by amki**
Adds triggers for p3n.

These are tested and correct as far as I'm concerned.

### CODE REVIEW

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    },
+    {
+      id: 'P3N Ashen Eye',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66AB', source: 'Sparkfledged' }),
+      alertText: (data, matches, output) => {
+        if (!data.ashenEyeDirections)
+          data.ashenEyeDirections = [];
+        switch (matches.heading) {
+          case '-3.14':
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p3n.ts:134]: Same comments as on P2N about not using exact floating point numbers for positions/headings and using a formula to round from radians to 0, 1, 2, 3 directions.

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+              first = output.e!();
+              break;
+            case 'south':
+              first = output.n!();
+              break;
+            case 'east':
+              first = output.w!();
+              break;
+          }
+          return first + ' then ' + safeSpot;
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p3n.ts:183]: Similar comment on P2N about using a `combo` style outputstring here.

I'd also shorten this to `X => Y` instead of using `then` as well.  TTS will transform that into "then" automatically, but it's easier to read (imo) on screen.

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+      id: 'P3N Darkened Fire Aoe',
+      type: 'HeadMarker',
+      netRegex: NetRegexes.headMarker({ id: ['010[C-F]'] }),
+      condition: Conditions.targetIsYou(),
+      alertText: (_data, _matches, output) => {
+        return output.text!();
+      },
+      outputStrings: {
+        text: {
+          en: 'Go to add',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p3n.ts:64]: Maybe `Stand on Darkened Fire`?

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    {
+      id: 'P3N Searing Breeze',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66B6', source: 'Phoinix', capture: false }),
+      response: Responses.stackThenSpread(),
+    },
+    {
+      id: 'P3N Spread Marker',
+      type: 'HeadMarker',
+      netRegex: NetRegexes.headMarker({ id: ['008B'] }),
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p3n.ts:217]:
```suggestion
      netRegex: NetRegexes.headMarker({ id: '008B' }),
```

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+      id: 'P3N Heat of Condemnation',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66B2', source: 'Phoinix' }),
+      suppressSeconds: 1,
+      response: Responses.tankBuster(),
+    },
+    {
+      id: 'P3N Darkened Fire Aoe',
+      type: 'HeadMarker',
+      netRegex: NetRegexes.headMarker({ id: ['010[C-F]'] }),
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p3n.ts:57]:
```suggestion
      netRegex: NetRegexes.headMarker({ id: '010[C-F]' }),
```

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+import ZoneId from '../../../../../resources/zone_id';
+import { RaidbossData } from '../../../../../types/data';
+import { TriggerSet } from '../../../../../types/trigger';
+
+export interface Data extends RaidbossData {
+  ashenEyeDirections?: string[];
+}
+
+const triggerSet: TriggerSet<Data> = {
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts:14]: No idea which ZoneId import you mean because line 5 has the ZoneID import I see in other files.

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  zoneId: ZoneId.AsphodelosTheThirdCircle,
+  timelineFile: 'p3n.txt',
```
[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Removed in 1be9b4e868f8ab0e3c7278e054e1fb9fb7df5be7

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Ironically it lets me commit without it just VS Code is stuck with that linter warning 🤔

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+      id: 'P3N Fore/Rear Carve',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66A[78]', source: 'Sunbird', capture: false }),
+      suppressSeconds: 1,
+      alertText: (_data, _matches, output) => {
+        return output.text!();
+      },
+      outputStrings: {
+        text: {
+          en: 'Dodge Add Cleave',
```
[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: I removed it in 1be9b4e868f8ab0e3c7278e054e1fb9fb7df5be7

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 3 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Changed in 1be9b4e868f8ab0e3c7278e054e1fb9fb7df5be7

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 4 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Changed in 1be9b4e868f8ab0e3c7278e054e1fb9fb7df5be7

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 5 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Changed in 1be9b4e868f8ab0e3c7278e054e1fb9fb7df5be7

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Yeah yeah, I meant that ZoneId import, exactly.  Huh! I don't know why VSCode is confused.  Sometimes I can get it to sort itself out if I close the file and reopen it?  At worst sometimes VSCode needs to be fully closed (if definitions change too much), but I wouldn't expect that to be needed here.

[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Wow you are right after closing and reopening the linter message is gone. I wouldn't have thought that it would have such a bug. In that case it's resolved as well.

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+      id: 'P3N Heat of Condemnation',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66B2', source: 'Phoinix' }),
+      suppressSeconds: 1,
+      response: Responses.tankBuster(),
+    },
+    {
+      id: 'P3N Darkened Fire Aoe',
+      type: 'HeadMarker',
+      netRegex: NetRegexes.headMarker({ id: '010[C-F]' }),
```
Comment (JLGarber) [inline ui/raidboss/data/06-ew/raid/p3n.ts:55]:
```suggestion
      netRegex: NetRegexes.headMarker({ id: ['010C', '010D', '010E', '010F'] }),
```

Apologies for a tiny nit on something you already updated, but where possible it might be nice to have IDs be individually searchable.

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66B8', source: 'Phoinix', capture: false }),
+      response: Responses.aoe(),
+    },
+    {
+      id: 'P3N Heat of Condemnation',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '66B2', source: 'Phoinix' }),
+      suppressSeconds: 1,
+      response: Responses.tankBuster(),
```
Comment (JLGarber) [inline ui/raidboss/data/06-ew/raid/p3n.ts:50]:
```suggestion
      response: Responses.tankCleave(),
```

At least one of my logs has an instance of this being a splash.

Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p3n.ts:50]: Maybe `tankCleave('alert')`? Maybe it should be alert by default to match buster, but it's not right now.

Comment (JLGarber) [reply ui/raidboss/data/06-ew/raid/p3n.ts:50]: Ah, that’s a good point, yes. I think you’re right that it should be `alert` by default.

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 9 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts:55]: I was orienting on triggers like https://github.com/quisquous/cactbot/blob/main/ui/raidboss/data/05-shb/raid/e8s.ts#L152 which are searchable with the same method as this one would be

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+          ko: '돌진을 피하세요',
+        },
+      },
+    },
+    {
+      id: 'P3N Sunbird Spawn',
+      type: 'AddedCombatant',
+      netRegex: NetRegexes.addedCombatant({ name: 'Sunbird', capture: false }),
+      suppressSeconds: 1,
+      response: Responses.killAdds(),
```
[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: I did an explaning callout for tanks in e32226eea0497fd9f12913a9d7136930d24c0a28 (because who else is going to split the adds realistically?) is that one ok?

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 1 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Fixed in e32226eea0497fd9f12913a9d7136930d24c0a28, still amazing formula.

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 2 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Fixed in e32226eea0497fd9f12913a9d7136930d24c0a28

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 10 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Fixed in e32226eea0497fd9f12913a9d7136930d24c0a28

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Akurosia) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: in t12 there was just a callout that the add is comming soon without further information.
It should be a "Kill adds without overlapping dead bodies" so 1 could be dead in the center and dont actually have to be on the edge

[the remark this answers is not in the document]
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Pretty sure it's not about the actual models stacking but about the circles not intersecting, I could be wrong though the one time I wiped to this they were stacked on point.

File: ui/raidboss/data/06-ew/raid/p3n.ts [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+      netRegex: NetRegexes.addedCombatant({ name: 'Sunbird', capture: false }),
+      suppressSeconds: 1,
+      alertText: (data, _matches, output) => {
+        if (data.role === 'tank')
+          return output.tank!();
+        return output.text!();
+      },
+      outputStrings: {
+        tank: {
+          en: 'Split Adds Don\'t Stack Circles',
```
Comment (quisquous) [inline ui/raidboss/data/06-ew/raid/p3n.ts:109]: What about:

```suggestion
          en: 'Pull add circles apart',
```

**Comment by quisquous** [comment]
The p2n merge conflicts with this.  Could you merge/rebase against main?

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 12 above)
Comment (amki) [reply ui/raidboss/data/06-ew/raid/p3n.ts]: Sure, whatever lingo you prefer :) changed in ce1618196bf3bd366946d76a4e8a61c8dde9aeab

**Comment by amki** [comment]
The conflict is with raidboss_manifest.txt which will always happen when appending lines one by one. Is there a way to avoid this I'm not thinking about?

Regardless I rebased it and it should apply now.

**Comment by quisquous** [comment]
> The conflict is with raidboss_manifest.txt which will always happen when appending lines one by one. Is there a way to avoid this I'm not thinking about?

There's not.  If you have two PRs, you can one PR1 say its target is main and the PR2 say its target is PR1, but github is not particularly smart about handling this scenario once PR1 is merged and often needs a rebase.  git also has "merge drivers" which run custom merge scripts when there are conflicts, but github doesn't want to run custom code during merges I guess and doesn't support this.

File: ui/raidboss/data/06-ew/raid/p3n.ts (hunk 9 above)
Comment (quisquous) [reply ui/raidboss/data/06-ew/raid/p3n.ts:55]: Oops, sorry, I forgot that this was still unresolved.  Agreed that E8S doesn't do this, but I think JLGarber was suggesting that in the future it makes life easier for developers to list out ids in full so that grep/ctrl+f style searches for full ids will turn them up more easily.

Comment (JLGarber) [reply ui/raidboss/data/06-ew/raid/p3n.ts:55]: Sorry, yes, I should have worded that a lot more clearly. It's a minimal cost now to write them in expanded form, while it could easily be more difficult to search for them in regex form later. Obviously not worth changing here, but definitely for the future.
